federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java

```diff
@@ -450,12 +523,20 @@ protected List<LDAPObject> searchLDAP(RealmModel realm, Map<String, String> attr
                 Condition usernameCondition = conditionsBuilder.equal(UserModel.USERNAME, attributes.get(UserModel.USERNAME), EscapeStrategy.NON_ASCII_CHARS_ONLY);
                 ldapQuery.addWhereCondition(usernameCondition);
 
-                List<LDAPObject> ldapObjects = ldapQuery.getResultList();
+                //need to handle offset manually as there has to be results collected first due to possible duplicate resutls from different fields
+                List<LDAPObject> ldapObjects = paginatedSearchLDAP(ldapQuery, null, ldapMaxResult);
+                
                 results.addAll(ldapObjects);
                 unique.addAll(ldapObjects);
             }
         }
 
+        // if ldapMaxResult == null -> no pagination
+        // if result has already enough values return
+        if (ldapMaxResult != null && ldapMaxResult.compareTo(results.size()) <= 0) {
+            return paginateResults(results, firstResult, maxResults);
```

Comment:
Can we change the method to return Stream and user `paginatedStream` method so we are consistent?

Reply:
Yes, I've had the `Stream` returned in one of my previous version, then I decided to switch it back due to a fact that it required concatenation of Streams from different calls (line 383) and more complex handling of `firstResult` and `maxResult` (as it's not easy to determine number of elements in the stream - line 380). But I may be missing something. 

Also I have a feeling that there is some issue with not-closing streams when concatenating (something with Spliterators) , but it might be unrelated to it.

Reply:
Changed to `Stream`.